File: Archive#00-001-100/042-Trapping Rain Water/leetcode042.py
class Solution:
    def trap(self, height: List[int]) -> int:
        ans, cur, stk = 0, 0, []
        while cur < len(height):
            while len(stk) != 0 and height[cur] > height[stk[-1]]:
                top, stk = stk[-1], stk[:-1]
                if len(stk) == 0: break
                width = cur - stk[-1] - 1
                depth = min(height[cur], height[stk[-1]]) - height[top]
                ans += width * depth
            stk.append(cur)
            cur += 1
        return ans

File: Archive#00-001-100/042-Trapping Rain Water/leetcode042.rb
# @param {Integer[]} height
# @return {Integer}
def trap(height)
  ans, cur, stk = 0, 0, []
  while cur < height.length do
    while stk.length > 0 && height[cur] > height[stk[-1]] do
      top = stk.delete_at(-1)
      if stk.length == 0 then
        break
      end
      width = cur - stk[-1] - 1
      depth = [height[cur], height[stk[-1]]].min - height[top]
      ans += width * depth
    end
    stk << cur
    cur += 1
  end
  ans
end